src/main/java/com/fasterxml/jackson/databind/jsontype/impl/TypeDeserializerBase.java

```diff
@@ -42,6 +43,8 @@ public abstract class TypeDeserializerBase
      */
     protected final JavaType _defaultImpl;
 
+    protected final ReentrantLock _defaultImplLock = new ReentrantLock();
```

Comment:
I don't think `JavaType` would be the place: locking here is specifically to try to avoid duplicate lookups for deserializer used for "default type" (of given polymorphic base type).

However. Since this is not done for correctness but as (possibly misguided?) performance optimization, maybe even use of basic `AtomicReference` would work (accepting possible race condition).